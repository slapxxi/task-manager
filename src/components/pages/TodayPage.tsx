import { createTag, matchTasksByTags } from '@lib';
import { TasksEditor } from '@local/components';
import IconSystem from '@local/components/IconSystem/IconSystem';
import Tag from '@local/components/Tags/Tag';
import Tags from '@local/components/Tags/Tags';
import { useStore } from '@local/hooks';
import { Tag as ITag } from '@local/types';
import { isToday } from 'date-fns';
import flatten from 'lodash-es/flatten';
import includes from 'lodash-es/includes';
import sortBy from 'lodash-es/sortBy';
import uniqBy from 'lodash-es/uniqBy';
import React, { useState } from 'react';
import PageTitle from './PageTitle';
import styles from './styles.css';

function TodayPage() {
  const { tasks, actions } = useStore();
  const [selectedTags, setSelectedTags] = useState<ITag[]>([]);
  const todayTasks = tasks.filter((t) => t.deadline && isToday(t.deadline));
  const todayTags = sortBy(
    uniqBy(flatten(todayTasks.map((t) => t.tags)), 'id') as ITag[],
    'name',
  );
  const matchingTasks = matchTasksByTags(todayTasks, selectedTags);

  function handleSelectTag(isSelected: boolean, tag: ITag) {
    if (isSelected) {
      if (tag.id === 'all') {
        selectAllTags();
        return;
      }
      selectTag(tag);
      return;
    } else {
      if (tag.id === 'all') {
        deselectAllTags();
        return;
      }
      deselectTag(tag);
    }
  }

  function selectAllTags() {
    setSelectedTags(todayTags);
  }

  function selectTag(tag: ITag) {
    setSelectedTags([...selectedTags, tag]);
  }

  function deselectAllTags() {
    setSelectedTags([]);
  }

  function deselectTag(tag: ITag) {
    setSelectedTags(selectedTags.filter((t) => t.id !== tag.id));
  }

  return (
    <div className={styles.todayPage}>
      <header className={styles.header}>
        <IconSystem name="star" size={25} />
        <PageTitle className={styles.title}>Today</PageTitle>
        <IconSystem name="dots" size={20} className={styles.menu} />
      </header>
      <div className={styles.tags}>
        <Tags
          tags={[createTag({ id: 'all', name: 'All' }), ...todayTags]}
          renderTag={({ tag }) => (
            <Tag
              tag={tag}
              parentId="today-page"
              onSelect={handleSelectTag}
              selected={
                tag.id === 'all'
                  ? selectedTags.length === todayTags.length
                  : includes(selectedTags.map((t) => t.id), tag.id)
              }
            />
          )}
        />
      </div>
      <TasksEditor
        tasks={matchingTasks}
        onEdit={actions.updateTask}
        onDelete={actions.deleteTask}
      />
    </div>
  );
}

export default TodayPage;
