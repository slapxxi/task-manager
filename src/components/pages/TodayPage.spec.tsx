import React from 'react';
import { MemoryRouter as Router } from 'react-router-dom';
import { render } from 'react-testing-library';
import TodayPage from './TodayPage';

jest.mock('@local/components/IconSystem/IconSystem');

it('renders', () => {
  const { container } = render(
    <Router>
      <TodayPage />
    </Router>,
  );
  expect(container.firstChild).toMatchSnapshot();
});
