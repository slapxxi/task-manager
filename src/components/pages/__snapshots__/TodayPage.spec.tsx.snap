// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<div
  class="todayPage"
>
  <div
    class="toolbar"
  >
    <ul
      class="container"
    >
      <li
        class="item"
      >
        <a
          aria-current="page"
          class="link active"
          href="/"
        >
          <div
            name="inbox"
            size="20"
          >
            icon-
            inbox
          </div>
        </a>
      </li>
    </ul>
  </div>
  <header
    class="header"
  >
    <div
      name="star"
      size="25"
    >
      icon-
      star
    </div>
    <h1
      class="title"
    >
      Today
    </h1>
    <div
      class="menu"
      name="dots"
      size="20"
    >
      icon-
      dots
    </div>
  </header>
  <div
    class="tags"
  >
    <ul
      class="tags"
      data-testid="tags"
    >
      <li
        class="tag"
      >
        <input
          checked=""
          class="checkbox"
          data-testid="checkbox"
          id="today-page-all"
          type="checkbox"
        />
        <label
          class="label"
          for="today-page-all"
        >
          All
        </label>
      </li>
    </ul>
  </div>
  <div
    class="content"
  >
    <p>
      Tasks are empty.
    </p>
    <footer
      class="footer"
    />
  </div>
</div>
`;
